presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/CommitCleaner.java

```diff
@@ -0,0 +1,233 @@
+                .map(shard -> createJdbi(shard.getConnection()))
+                .map(dbi -> dbi.onDemand(ShardCommitCleanerDao.class))
+                .collect(toImmutableList());
+
+        this.clock = requireNonNull(clock, "clock is null");
+    }
+
+    public synchronized void removeOldCommits()
+    {
+        long activeCommitId = transactionManager.oldestActiveCommitId();
+
+        Set<Long> droppedTableIds = masterDao.getDroppedTableIds(activeCommitId);
+
+        Map<Integer, Set<Long>> droppedTableIdsByShard = droppedTableIds.stream()
+                .collect(groupingBy(tableId -> tableShard(tableId, shardDao.size()), toSet()));
+
+        // drop index tables
+        for (long tableId : droppedTableIds) {
+            shardDao.get(tableShard(tableId, shardDao.size())).useHandle(handle ->
+                    handle.execute("DROP TABLE IF EXISTS " + chunkIndexTable(tableId)));
+        }
+
+        // cleanup deleted chunks
+        long now = clock.millis();
+        for (int i = 0; i < shardDao.size(); i++) {
+            ShardCommitCleanerDao shard = shardDao.get(i);
+            Set<Long> tableIds = droppedTableIdsByShard.getOrDefault(i, ImmutableSet.of());
+
+            // cleanup chunks for old commits
+            while (true) {
+                List<TableChunk> chunks = shard.getDeletedChunks(activeCommitId, CHUNK_BATCH_SIZE);
+                if (chunks.isEmpty()) {
+                    break;
+                }
+                cleanupChunks(shard, now, chunks, tableIds);
+            }
+
+            // cleanup chunks for dropped tables
+            if (!tableIds.isEmpty()) {
+                while (true) {
+                    List<TableChunk> chunks = shard.getDeletedChunks(tableIds, CHUNK_BATCH_SIZE);
+                    if (chunks.isEmpty()) {
+                        break;
+                    }
+                    cleanupChunks(shard, now, chunks, tableIds);
+                }
+            }
+        }
+
+        // cleanup dropped table columns
+        if (!droppedTableIds.isEmpty()) {
+            masterDao.cleanupDroppedTableColumns(droppedTableIds);
+        }
+
+        // cleanup schemas, tables, columns, views, commits
+        masterDao.cleanup(activeCommitId);
+
+        // cleanup table sizes
+        for (ShardCommitCleanerDao dao : shardDao) {
+            dao.cleanupTableSizes(activeCommitId);
+        }
+
+        // cleanup successful transactions
+        Set<Long> transactionIds = masterDao.getSuccessfulTransactionIds();
+        for (Iterable<Long> ids : partition(transactionIds, 1000)) {
+            for (ShardCommitCleanerDao shard : shardDao) {
+                shard.cleanupCreatedChunks(ids);
+            }
+            masterDao.cleanupTransactions(ids);
+        }
+
+        // abort failed transactions
+        masterDao.abortTransactions(transactionManager.activeTransactions(), clock.millis());
+
+        // cleanup failed transactions
+        now = clock.millis();
```

Comment:
Is there a reason we need different methods to get time?
clock.millis(), and  System.currentTimeMillis()

Shall we migrate to clock.millis() in every place? not saying now, but wanna know what's the right thing.

Reply:
Good observation. We use `Clock` in this class because we need to inject a time source for testing. We could also inject it in other places, but that seems needlessly verbose.